### ISSUE TITLE
Pre-processing Pipeline

### ISSUE THREAD

**Issue by chenqi-li** [Edoar-do/HuBERT-ECG#1]
Hi,

Thank you for the interesting work and open-sourcing the code & pre-trained weights. 

I would like to try your pre-trained model on a different dataset. What format/structure should I pre-process my data? Alternatively, if possible, could you please provide the pre-processing for PTB dataset?

Thanks.

**Comment by Edoar-do** [comment]
Hi @chenqi-li !
Thank you for your interest in our work!
HuBERT-ECG accepts 5-second 12-leads ECGs where the leads are concatenated with each other following this order: I, II, III, aVR, aVL, aVF, V1, ... V6. The resulting input tensor has shape (batch_size, 5 x 100 x 12).
If the ECGs you have last less than 5 seconds, you can pad them and adjust the `attention_mask` accordingly. A simple implementation for its computation is provided in `dataset.py`.

**Extra tip**: if your ECGs last more than 5 seconds, we suggest you to use a time-aligned random-crop as data-augmentation (see `dataset.py`). This simple augmentation helped us reach much better downstream performance

PTB is one of the datasets onto which we evaluated HuBERT-ECG, so you might want to take a look at the preprint ;) 

Should you have any suggestion, please let us know.

Happy coding!

**Comment by chenqi-li** [comment]
Thanks @Edoar-do for the timely response and clarification!

According to the pre-print, for pre-processing, you applied bandpass filter to [0.05, 47]Hz, downsample to 100Hz and rescale to [-1, 1] range similar to Natarajan et al. Since I want to use your pre-trained weights, I thought it would be best for me to apply the same pre-processing to my dataset, but I cannot seem to find the code that does the filtering and rescale to [-1, 1].

I am trying to replicate the same pre-processing pipeline for my dataset and have a few questions:
1. Could you please clarify how you did the rescaling? It is quite common to do channel-wise z-score normalization, which gives you zero mean and unit variance, but not all data would strictly lie within [-1, 1]. Do I have to make sure that every data-point is strictly within the [-1, 1] range? Also, in Natarajan et al, they applied that on a per-channel basis, i.e. should every channel have min -1, max 1? 
2. To ensure consistency, which library did you use to implement the FIR filters?
3. Are there any package requirements? e.g. required PyTorch, Python versions?

Regarding the input size, I have 30s segments and would like to extract features/representations using HuBERT-ECG, as opposed to making downstream predictions. Since HuBERT-ECG accepts 5s inputs, I am thinking to divide the 30s segment into six segments of 5s each and concatenate/average the six feature/representation vectors. Do you have any suggestions for this?

Thanks in advance!

**Comment by Edoar-do** [comment]
@chenqi-li  Oops, I just updated `utils.py` with ecg pre-processing code. Sorry about that :)

Quick note: downsampling to 100 hz is performed on-the-fly when yielding ECGs before batching. This choice was necessary as we experimented with multiple downsampling factors before finding that downsampling to 100 hz was optimal for us.

Concerning your use case, both averaging and concatenation seem valid approaches worth trying, but I'd prefer concatenation because, if you are dealing with conditions whose patterns can manifest only in certain (unknown a-priori) frames of the ecg, averaging might lead to the "erosion" of the target information. In contrast, concatenation sounds as a more robust approach that preserves all the information content.
